### PR TITLE
upgrade crds api version to a stable version

### PR DESCRIPTION
**Pull request by zhu733756**
Signed-off-by: zhu733756 <talonzhu@yunify.com>

**What type of PR is this?**
Upgrade crds api version to a stable version.
/kind feature

**What this PR does / why we need it**:
Upgrade crds api version to a stable version for  better management.

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubeedge/kubeedge/issues/2780

**Special notes for your reviewer**:

Kubebuilder relies on annotations to build crds, and in this case, controllers don't care about version(each component has a single version). Therefore, I replace the version information by regex matches. The related filenames would modify containing v1 if this pr is ok.

/cc @fisherxu

### CODE REVIEW

**Review by fisherxu — COMMENTED** [review]
Maybe we also need some adjust work for the crd content since the e2e tests failed.

**Comment by zhu733756** [comment]
> Maybe we also need some adjust work for the crd content since the e2e tests failed.

Yeah, I see. I will try to make a shell script to convert it. Do you have other idea?

**Comment by fisherxu** [comment]
> > Maybe we also need some adjust work for the crd content since the e2e tests failed.
> 
> Yeah, I see. I will try to make a shell script to convert it.

https://github.com/kubeedge/kubeedge/pull/2408  Have a pending pr to do this, can have a look :) 

>  Do you have other idea?

Have no other ideas, just about kubebuilder.

**Comment by zhu733756** [comment]
Thanks. I will have a look at it.

**Comment by fisherxu** [comment]
/reopen

#2408 have been pending for a long time, you can do it in this PR, thanks :)

**Comment by zhu733756** [comment]
> /reopen
> 
> #2408 has been pending for a long time, you can do it in this PR, thanks :)

Updated, pls have a review on it, thanks.

**Comment by fisherxu** [comment]
> How about change the api version of the crds to apiextensions.k8s.io/v1 ?

@zhu733756 Change crds files to apiextensions.k8s.io/v1, but no need to upgrade the device version. Upgrade device to v1 means that they will be GA, Now we still in alpha status.

File: build/crd-samples/devices/CC2650-device-instance.yaml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 
-apiVersion: devices.kubeedge.io/v1alpha2
```
Comment (fisherxu) [inline build/crd-samples/devices/CC2650-device-instance.yaml:2]: Please revert these changes, thanks.

Comment (zhu733756) [reply build/crd-samples/devices/CC2650-device-instance.yaml:2]: ok

Comment (zhu733756) [reply build/crd-samples/devices/CC2650-device-instance.yaml:2]: The request is restricted by github，i will push later.

**Review by fisherxu — COMMENTED** [review]
Looks like the e2e is still failed

**Comment by zhu733756** [comment]
Yeah, I see.
I just apply the update from the crds. I found all other tests passed, execpt for the test for CustomizedDeviceInstance method. The error message is displayed as follows:
```
The device is not matching with what was expected:
-----
Expected:
TypeMeta:{Device devices.kubeedge.io/v1alpha2},
Namespace:default, 
Labels:map[description:TISimplelinkSensorTag manufacturer:TexasInstruments model:CC2650], 
Spec:{&LocalObjectReference{Name:sensor-tag-customized-model,} {%!s(*v1alpha2.ProtocolConfigOpcUA=<nil>) %!s(*v1alpha2.ProtocolConfigModbus=<nil>) %!s(*v1alpha2.ProtocolConfigBluetooth=<nil>) %!s(*v1alpha2.ProtocolConfigCommon=&{0xc000316580 <nil>  0 0 0 0  <nil>}) %!s(*v1alpha2.ProtocolConfigCustomized=&{CustomizedProtocol1 0xc0000102c8})} [{temperature %!s(int64=0) %!s(int64=0) %!s(*v1alpha2.CustomizedValue=<nil>) {%!s(*v1alpha2.VisitorConfigOPCUA=<nil>) %!s(*v1alpha2.VisitorConfigModbus=<nil>) %!s(*v1alpha2.VisitorConfigBluetooth=<nil>) %!s(*v1alpha2.VisitorConfigCustomized=&{CustomizedProtocol1 `0xc0000102c8`})}} {temperature-enable %!s(int64=0) %!s(int64=0) %!s(*v1alpha2.CustomizedValue=<nil>) {%!s(*v1alpha2.VisitorConfigOPCUA=<nil>) %!s(*v1alpha2.VisitorConfigModbus=<nil>) %!s(*v1alpha2.VisitorConfigBluetooth=<nil>) %!s(*v1alpha2.VisitorConfigCustomized=&{CustomizedProtocol1 `0xc0000102e0`})}}] {[] } &NodeSelector{NodeSelectorTerms:[]NodeSelectorTerm{NodeSelectorTerm{MatchExpressions:[]NodeSelectorRequirement{NodeSelectorRequirement{Key:,Operator:In,Values:[edge-node],},},MatchFields:[]NodeSelectorRequirement{},},},}}
----
Got:
TypeMeta:{Device devices.kubeedge.io/v1alpha2}, 
Namespace:default,
Labels:map[description:TISimplelinkSensorTag manufacturer:TexasInstruments model:CC2650], 
Spec:{&LocalObjectReference{Name:sensor-tag-customized-model,} {%!s(*v1alpha2.ProtocolConfigOpcUA=<nil>) %!s(*v1alpha2.ProtocolConfigModbus=<nil>) %!s(*v1alpha2.ProtocolConfigBluetooth=<nil>) %!s(*v1alpha2.ProtocolConfigCommon=&{0xc000316840 <nil>  0 0 0 0  <nil>}) %!s(*v1alpha2.ProtocolConfigCustomized=&{CustomizedProtocol1 0xc000010f70})} [{temperature %!s(int64=0) %!s(int64=0) %!s(*v1alpha2.CustomizedValue=<nil>) {%!s(*v1alpha2.VisitorConfigOPCUA=<nil>) %!s(*v1alpha2.VisitorConfigModbus=<nil>) %!s(*v1alpha2.VisitorConfigBluetooth=<nil>) %!s(*v1alpha2.VisitorConfigCustomized=&{CustomizedProtocol1 `0xc000010f10`})}} {temperature-enable %!s(int64=0) %!s(int64=0) %!s(*v1alpha2.CustomizedValue=<nil>) {%!s(*v1alpha2.VisitorConfigOPCUA=<nil>) %!s(*v1alpha2.VisitorConfigModbus=<nil>) %!s(*v1alpha2.VisitorConfigBluetooth=<nil>) %!s(*v1alpha2.VisitorConfigCustomized=&{CustomizedProtocol1 `0xc000010f68`})}}] {[] } &NodeSelector{NodeSelectorTerms:[]NodeSelectorTerm{NodeSelectorTerm{MatchExpressions:[]NodeSelectorRequirement{NodeSelectorRequirement{Key:,Operator:In,Values:[edge-node],},},MatchFields:[]NodeSelectorRequirement{},},},}}\n
```
The difference is that the configmap data is different. Can you help to solve this issue?

> Looks like the e2e is still failed

**Comment by zhu733756** [comment]
try to fix it.

- refers to https://github.com/kubernetes-sigs/controller-tools/issues/554, it suggests us to use  json.RawMessage, instead of interface{}, however, this method is too destructive.
- refers to  https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/

> Yeah, I see.
> I just apply the update from the crds. I found all other tests passed, execpt for the test for CustomizedDeviceInstance method. The error message is displayed as follows:
> 
> ```
> The device is not matching with what was expected:
> -----
> expected：
> {\"deviceModelRef\":{\"name\":\"sensor-tag-customized-model\"},\"protocol\":{\"common\":{\"com\":{\"serialPort\":\"/dev/ttyS0\",\"baudRate\":9600,\"dataBits\":8,\"parity\":\"even\",\"stopBits\":1}},\"customizedProtocol\":{\"protocolName\":\"CustomizedProtocol1\",\"configData\":{\"config1\":\"config-val1\",\"config2\":\"config-val2\"}}},\"propertyVisitors\":[{\"propertyName\":\"temperature\",\"customizedProtocol\":{\"protocolName\":\"CustomizedProtocol1\",\"configData\":{\"config1\":\"config-val1\",\"config2\":\"config-val2\"}}},{\"propertyName\":\"temperature-enable\",\"customizedProtocol\":{\"protocolName\":\"CustomizedProtocol1\",\"configData\":{\"config3\":\"config-val3\",\"config4\":\"config-val4\"}}}],\"data\":{},\"nodeSelector\":{\"nodeSelectorTerms\":[{\"matchExpressions\":[{\"key\":\"\",\"operator\":\"In\",\"values\":[\"edge-node\"]}]}]}}
> ----
> got:
> {\"deviceModelRef\":{\"name\":\"sensor-tag-customized-model\"},\"protocol\":{\"common\":{\"com\":{\"serialPort\":\"/dev/ttyS0\",\"baudRate\":9600,\"dataBits\":8,\"parity\":\"even\",\"stopBits\":1}},\"customizedProtocol\":{\"protocolName\":\"CustomizedProtocol1\",\"configData\":{}}},\"propertyVisitors\":[{\"propertyName\":\"temperature\",\"customizedProtocol\":{\"protocolName\":\"CustomizedProtocol1\",\"configData\":{}}},{\"propertyName\":\"temperature-enable\",\"customizedProtocol\":{\"protocolName\":\"CustomizedProtocol1\",\"configData\":{}}}],\"data\":{},\"nodeSelector\":{\"nodeSelectorTerms\":[{\"matchExpressions\":[{\"key\":\"\",\"operator\":\"In\",\"values\":[\"edge-node\"]}]}]}}
> ```
> 
> The difference is that the configmap data is different. Can you help to solve this issue?
> 
> > Looks like the e2e is still failed

**Comment by fisherxu** [comment]
/lgtm
Is the new crd files generated by any tools? @zhu733756

**Comment by zhu733756** [comment]
> /lgtm
> Are the new crd files generated by any tools? @zhu733756

Based on your recommendation, I used KubeBuilder to generate CRD, and made some field modifications according to the official documents.

**Comment by fisherxu** [comment]
> > /lgtm
> > Are the new crd files generated by any tools? @zhu733756
> 
> Based on your recommendation, I used KubeBuilder to generate CRD, and made some field modifications according to the official documents.


Okey, let's use KubeBuilder directly in future and can add a verify test in the ci.

/approve
